internal/controllers/storage/sync.go

```diff
@@ -354,5 +373,84 @@ func (r *Reconciler) getSecretKey(
 			secretKeyRef.Name,
 		)
 	}
+
 	return string(secretVal), nil
 }
+
+func (r *Reconciler) handlePauseResume(
+	ctx context.Context,
+	storage *resources.StorageClusterBuilder,
+) (bool, ctrl.Result, error) {
+	r.Log.Info("running step handlePauseResume for Storage")
+	if storage.Status.State == StorageReady && storage.Spec.Pause == PausedState {
+		r.Log.Info("`pause: Paused` was noticed, attempting to delete Storage StatefulSet")
+		statefulSet := &appsv1.StatefulSet{}
+		err := r.Client.Get(ctx,
+			types.NamespacedName{
+				Name:      storage.Name,
+				Namespace: storage.Namespace,
+			},
+			statefulSet,
+		)
+		if err != nil {
+			r.Log.Error(err, "failed to get the Storage StatefulSet object before deletion")
+			return Stop, ctrl.Result{RequeueAfter: DefaultRequeueDelay}, err
+		}
+
+		err = r.Client.Delete(ctx, statefulSet)
+		if err != nil {
+			r.Log.Error(err, "failed to delete the Storage StatefulSet object when moving from Ready -> Paused")
+			return Stop, ctrl.Result{RequeueAfter: DefaultRequeueDelay}, err
+		}
+
+		meta.SetStatusCondition(&storage.Status.Conditions, metav1.Condition{
+			Type:    StoragePausedCondition,
+			Status:  "True",
+			Reason:  StoragePausedReason,
+			Message: "pause: Paused is set on Storage",
+		})
+
+		storage.Status.StateBeforePausing = storage.Status.State
+		storage.Status.State = StoragePaused
+
+		return r.setState(ctx, storage)
+	}
+
+	if storage.Status.State == StoragePaused && storage.Spec.Pause == RunningState {
+		r.Log.Info("`pause: Running` was noticed, moving Storage to `Pending`")
+		meta.RemoveStatusCondition(&storage.Status.Conditions, StoragePausedCondition)
+
+		storage.Status.State = storage.Status.StateBeforePausing
+		storage.Status.StateBeforePausing = ""
+		return r.setState(ctx, storage)
+	}
+
+	return Continue, ctrl.Result{}, nil
+}
+
+func (r *Reconciler) checkStorageFrozen(
+	ctx context.Context,
+	storage *resources.StorageClusterBuilder,
+) (bool, ctrl.Result, error) {
+	r.Log.Info("running step checkStorageFrozen for Storage")
+	if storage.Status.State != StorageFrozen && storage.Spec.Pause == FrozenState {
```

Comment:
I would argue it IS important for the user to see that his Storage\Database is Frozen. I thought about this state as just about a convenience feature for the user, and it's essentially free. Otherwise, the user might forget to un-set the Frozen parameter, then be surprised because operator has not reacted to any changes for the past N hours. :)

Reply:
I prefer using check in webhook to verify that we editing CR resources only when state `Frozen` set to `false`. That also prevent changes and should be clear to see. I don't like idea of saving previous state of object in `.status` field and trying to revert to this state when unfroze process. I believe that Reconcile and Sync function should be completely execute with minimum conditional operators